Source/Events.Processing/Projections/ProjectionsService.cs

```diff
@@ -0,0 +1,246 @@
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+                _logger.ProjectionDisconnected(arguments);
+            }
+        }
+
+        async Task<Try<IEnumerable<Task>>> TryStartProjection(
+            IReverseCallDispatcher<ProjectionClientToRuntimeMessage, ProjectionRuntimeToClientMessage, ProjectionRegistrationRequest, ProjectionRegistrationResponse, ProjectionRequest, ProjectionResponse> dispatcher,
+            ProjectionRegistrationArguments arguments,
+            StreamProcessor eventProcessorStreamProcessor,
+            CancellationToken cancellationToken)
+        {
+            _logger.StartingProjection(arguments);
+            try
+            {
+                var runningDispatcher = dispatcher.Accept(new ProjectionRegistrationResponse(), cancellationToken);
+
+                await ResetIfDefinitionChanged(arguments, cancellationToken).ConfigureAwait(false);
+                await eventProcessorStreamProcessor.Initialize().ConfigureAwait(false);
+                return new[] { eventProcessorStreamProcessor.Start(), runningDispatcher };
+            }
+            catch (Exception ex)
+            {
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    _logger.ErrorWhileStartingProjection(ex, arguments);
+                }
+
+                return ex;
+            }
+        }
+
+        Try<StreamProcessor> TryRegisterProjection(
+            ProjectionRegistrationArguments arguments,
+            Func<IEventProcessor> getEventProcessor,
+            CancellationToken cancellationToken)
+        {
+            _logger.RegisteringProjection(arguments);
+            try
+            {
+                return (_streamProcessors.TryRegister(
+                    arguments.ProjectionDefinition.Scope,
+                    arguments.ProjectionDefinition.Projection.Value,
+                    new EventLogStreamDefinition(),
+                    getEventProcessor,
+                    cancellationToken,
+                    out var outputtedEventProcessorStreamProcessor), outputtedEventProcessorStreamProcessor);
+            }
+            catch (Exception ex)
+            {
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    _logger.ErrorWhileRegisteringProjection(ex, arguments);
+                }
+
+                return ex;
+            }
+        }
+        async Task ResetIfDefinitionChanged(
+            ProjectionRegistrationArguments arguments,
+            CancellationToken token)
+        {
+            _logger.ComparingProjectionDefintion(arguments);
+            var tenantsAndComparisonResult = await _projectionDefinitionComparer.DiffersFromPersisted(arguments.ProjectionDefinition, token).ConfigureAwait(false);
+
+            await _onAllTenants.PerformAsync(async tenant =>
+            {
+                if (!tenantsAndComparisonResult.TryGetValue(tenant, out var comparisonResult)) return;
```

Comment:
It seems something terrible has gone wrong if the tenant is not in this dictionary?